### PR TITLE
Improvements in RequestParser

### PR DESCRIPTION
**Pull request by vladsud**
1. Perf - we keep recreating same object for same request
2. We incorrectly handle terminating requests in couple places, mostly due to not having easier to use tools (isLeaf)
3. sub-request code has extra dead code that can be eliminating by changing signature of the function.

### CODE REVIEW

File: packages/runtime/runtime-utils/src/requestParser.ts [hunk 1]
```diff
@@ -30,6 +30,13 @@ export class RequestParser implements IRequest {
     private requestPathParts: readonly string[] | undefined;
     public readonly query: string;
     constructor(private readonly request: Readonly<IRequest>) {
+        // Perf optimizations. Even better would be not to create this object
```
Comment (heliocliu) [inline packages/runtime/runtime-utils/src/requestParser.ts:33]: looks like `new RequestParser` is only called in a few places; if this is worth optimizing here, would it also be worth just adding a create method that does this check but spits back the same request under this condition?

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 2]
```diff
@@ -47,7 +47,7 @@ import {
     ISummaryTreeWithStats,
     CreateSummarizerNodeSource,
 } from "@fluidframework/runtime-definitions";
-import { generateHandleContextPath, SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import { generateHandleContextPath, SummaryTreeBuilder, RequestParser } from "@fluidframework/runtime-utils";
```
Comment (heliocliu) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:50]: nit: alphabetize (though not sure how much we really care since the linter doesn't care)

**Review by heliocliu — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts (hunk 2 above)
Comment (tylerbutler) [reply packages/runtime/datastore/src/dataStoreRuntime.ts]: Good catch. I didn't realize we didn't have the sorting lint rule enabled. I'll fix that.

**Review by anthony-murphy — COMMENTED** [review]
nice!

**Review by anthony-murphy — APPROVED** [review]

**Review by agarwal-navin — APPROVED** [review]

**Review by jatgarg — APPROVED** [review]

**Pull request merged**